executor/analyze.go

```diff
@@ -513,33 +545,484 @@ func analyzeFastExec(exec *AnalyzeFastExec) []analyzeResult {
+			e.sampTasks = append(e.sampTasks, sampTasksForRoutine[i]...)
+		}
+	}()
+
+	store, _ := e.ctx.GetStore().(tikv.Storage)
+	e.cache = store.GetRegionCache()
+	startKey, endKey := tablecodec.GetTableHandleKeyRange(e.physicalTableID)
+	// extract all regions contain the table
+	e.scanTasks = e.scanTasks[:0]
+	e.sampTasks = e.sampTasks[:0]
+	targetKey := startKey
+	for {
+		// Search for the region which contains the targetKey.
+		loc, err := e.cache.LocateKey(bo, targetKey)
+		if err != nil {
+			return false, err
+		}
+		if bytes.Compare(endKey, loc.StartKey) < 0 {
+			break
+		}
+		// Set the next search key.
+		targetKey = loc.EndKey
+
+		// If the KV pairs in the region all belonging to the table, add it to the sample task.
+		if bytes.Compare(startKey, loc.StartKey) <= 0 && bytes.Compare(loc.EndKey, endKey) <= 0 {
+			e.sampLocs <- loc
+			continue
+		}
+
+		e.scanTasks = append(e.scanTasks, loc)
+		if bytes.Compare(loc.StartKey, startKey) < 0 {
+			loc.StartKey = startKey
+		}
+		if bytes.Compare(endKey, loc.EndKey) < 0 || len(loc.EndKey) == 0 {
+			loc.EndKey = endKey
+			break
+		}
+	}
+
+	return false, nil
+}
+
+func (e *AnalyzeFastExec) decodeValues(sValue []byte) (values map[int64]types.Datum, err error) {
+	colID2FieldTypes := make(map[int64]*types.FieldType, len(e.colsInfo))
+	if e.pkInfo != nil {
+		colID2FieldTypes[e.pkInfo.ID] = &e.pkInfo.FieldType
+	}
+	for _, col := range e.colsInfo {
+		colID2FieldTypes[col.ID] = &col.FieldType
+	}
+	return tablecodec.DecodeRow(sValue, colID2FieldTypes, e.ctx.GetSessionVars().Location())
+}
+
+func (e *AnalyzeFastExec) getValueByInfo(colInfo *model.ColumnInfo, values map[int64]types.Datum) (types.Datum, error) {
+	val, ok := values[colInfo.ID]
+	if !ok {
+		return table.GetColOriginDefaultValue(e.ctx, colInfo)
+	}
+	return val, nil
+}
+
+func (e *AnalyzeFastExec) updateCollectorSamples(sValue []byte, sKey kv.Key, samplePos int32, hasPKInfo int) (err error) {
+	// Decode the cols value in order.
+	var values map[int64]types.Datum
+	values, err = e.decodeValues(sValue)
+	if err != nil {
+		return err
+	}
+	// Update the primary key collector.
+	if hasPKInfo > 0 {
+		v, ok := values[e.pkInfo.ID]
+		if !ok {
+			var key int64
+			_, key, err = tablecodec.DecodeRecordKey(sKey)
+			if err != nil {
+				return err
+			}
+			v = types.NewIntDatum(key)
+		}
+		if e.collectors[0].Samples[samplePos] == nil {
+			e.collectors[0].Samples[samplePos] = &statistics.SampleItem{}
+		}
+		e.collectors[0].Samples[samplePos].Ordinal = int(samplePos)
+		e.collectors[0].Samples[samplePos].Value = v
+	}
+	// Update the columns' collectors.
+	for j, colInfo := range e.colsInfo {
+		v, err := e.getValueByInfo(colInfo, values)
+		if err != nil {
+			return err
+		}
+		if e.collectors[hasPKInfo+j].Samples[samplePos] == nil {
+			e.collectors[hasPKInfo+j].Samples[samplePos] = &statistics.SampleItem{}
+		}
+		e.collectors[hasPKInfo+j].Samples[samplePos].Ordinal = int(samplePos)
```

Comment:
We cannot use `samplePos` as `Ordinal` here, `Ordinal` is used for computing order correlation of columns, so it should reflect the order of samples in the table, but here we are in a parallel context, so `samplePos` cannot guarantee the order now.